### PR TITLE
Adopt new tree in custom views

### LINKED ISSUE #63566: Adopt new tree for custom views
Adopt new tree for custom views

### PR DESCRIPTION
**Pull request by alexr00**
Fixes #63566

### CODE REVIEW

**Comment by alexr00** [comment]
Note: This is still missing auto expanding of nodes, a feature that doesn't yet exist in the new tree. Once the functionality is available I'll add it before merging.

File: src/vs/workbench/browser/parts/views/media/views.css [hunk 1]
```diff
@@ -93,19 +94,31 @@
...
+	background-size: 16px;
+	background-position: 50% 50%;
+	background-repeat: no-repeat;
+	padding-right: 6px;
+	width: 16px;
+	height: 22px;
+	display: inline-block;
+	vertical-align: top;
+	content: ' ';
+} */
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/media/views.css]: I think this commented rule can be deleted

File: src/vs/workbench/browser/parts/views/customView.ts [hunk 2]
```diff
@@ -530,136 +596,133 @@ export class CustomTreeView extends Disposable implements ITreeView {
...
-	renderTemplate(tree: ITree, templateId: string, container: HTMLElement): ITreeExplorerTemplateData {
+	renderTemplate(container: HTMLElement): ITreeExplorerTemplateData {
 		DOM.addClass(container, 'custom-view-tree-node-item');
 
 		const icon = DOM.append(container, DOM.$('.custom-view-tree-node-item-icon'));
-		const resourceLabel = this.labels.create(container, { supportHighlights: true, donotSupportOcticons: true });
-		DOM.addClass(resourceLabel.element, 'custom-view-tree-node-item-resourceLabel');
-		const actionsContainer = DOM.append(resourceLabel.element, DOM.$('.actions'));
+
+		const labelAndActionContainer = document.createElement('div');
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/customView.ts]: May I know why this new container for label and action is needed? Is the structure can't be as before?

File: src/vs/workbench/browser/parts/views/customView.ts [hunk 3]
```diff
@@ -669,17 +732,21 @@ class TreeRenderer implements IRenderer {
...
 		DOM.toggleClass(templateData.icon, 'custom-view-tree-node-item-icon', !!iconUrl);
 		templateData.actionBar.context = (<TreeViewItemHandleArg>{ $treeViewId: this.treeViewId, $treeItemHandle: node.handle });
 		templateData.actionBar.push(this.menus.getResourceActions(node), { icon: true, label: false });
+		this.setAlignment(templateData.container, node);
+		this._register(this.themeService.onDidFileIconThemeChange(() => this.setAlignment(templateData.container, node)));
+	}
 
-		templateData.aligner.treeItem = node;
+	private setAlignment(container: HTMLElement, treeItem: ITreeItem) {
+		DOM.toggleClass(container.parentElement!, 'align-icon-with-twisty', this.aligner.alignIconWithTwisty(treeItem));
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/customView.ts]: Does the class `align-icon-with-twisty` has to be set on the container?

File: src/vs/workbench/browser/parts/views/customView.ts [hunk 4]
```diff
@@ -530,136 +596,133 @@ export class CustomTreeView extends Disposable implements ITreeView {
...
 	}
 	const codeBackground = theme.getColor(textCodeBlockBackground);
 	if (codeBackground) {
 		collector.addRule(`.tree-explorer-viewlet-tree-view > .message code { background-color: ${codeBackground}; }`);
 	}
 });
 
-class TreeRenderer implements IRenderer {
+interface ITreeExplorerTemplateData {
+	elementDisposable: IDisposable;
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/customView.ts]: what is this `elementDisposable`?

File: src/vs/workbench/browser/parts/views/customView.ts [hunk 5]
```diff
@@ -530,136 +596,133 @@ export class CustomTreeView extends Disposable implements ITreeView {
...
+	container: HTMLElement;
+	resourceLabel: IResourceLabel;
+	icon: HTMLElement;
+	actionBar: ActionBar;
+	aligner: Aligner;
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/customView.ts]: No need to pass `aligner` through template.

File: src/vs/workbench/browser/parts/views/customView.ts [hunk 6]
```diff
@@ -669,17 +732,21 @@ class TreeRenderer implements IRenderer {
...
-			templateData.resourceLabel.setResource({ name: label, description }, { title, hideIcon: true, extraClasses: ['custom-view-tree-node-item-resourceLabel'], matches });
+			templateData.resourceLabel.setResource({ name: label, description }, { title, hideIcon: true, extraClasses: ['custom-view-tree-node-item-resourceLabel'], matches: createMatches(element.filterData) });
 		}
 
 		templateData.icon.style.backgroundImage = iconUrl ? `url('${iconUrl.toString(true)}')` : '';
 		DOM.toggleClass(templateData.icon, 'custom-view-tree-node-item-icon', !!iconUrl);
 		templateData.actionBar.context = (<TreeViewItemHandleArg>{ $treeViewId: this.treeViewId, $treeItemHandle: node.handle });
 		templateData.actionBar.push(this.menus.getResourceActions(node), { icon: true, label: false });
+		this.setAlignment(templateData.container, node);
+		this._register(this.themeService.onDidFileIconThemeChange(() => this.setAlignment(templateData.container, node)));
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/customView.ts]: This will dispose this listener only when the class is disposed. But we need to dispose this when the corresponding template is disposed. Hence, I would introduce a new property called `itemAligner` which is of type class `ItemAligner`. This new class is very similar to the `Aligner` but it aligns the given item. 

```
interface ITreeExplorerTemplateData {
	itemAligner: ItemAligner;
}
```

```
class ItemAligner extends Disposable {

	private _treeItem: ITreeItem;

	constructor(private readonly container: HTMLElement, private readonly aligner: Aligner) {
		this._register(this.themeService.onDidFileIconThemeChange(() => this.render()));
	}

	set treeItem(treeItem: ITreeItem) {
		this._treeItem = treeItem;
		this.render();
	}

	private render(): void {
		if (this._treeItem) {
			DOM.toggleClass(this.container, 'align-icon-with-twisty', this.aligner.alignIconWithTwisty(this._treeItem));
		}
	}	
}
```

And this `ItemAligner` can be disposed in `disposeTemplate`

**Review by sandy081 — CHANGES_REQUESTED** [review]
@alexr00 Changes look good and there are some minor suggestions. Please take a look. Thanks.

File: src/vs/workbench/browser/parts/views/customView.ts (hunk 3 above)
Comment (alexr00) [reply src/vs/workbench/browser/parts/views/customView.ts]: For the styles to apply correctly it must be on the container.

File: src/vs/workbench/browser/parts/views/customView.ts (hunk 2 above)
Comment (alexr00) [reply src/vs/workbench/browser/parts/views/customView.ts]: Left over from trying to fix something else. Removed.

File: src/vs/workbench/browser/parts/views/customView.ts (hunk 4 above)
Comment (alexr00) [reply src/vs/workbench/browser/parts/views/customView.ts]: Copied from other trees, but it is not needed here. removed.

File: src/vs/workbench/browser/parts/views/customView.ts (hunk 6 above)
Comment (alexr00) [reply src/vs/workbench/browser/parts/views/customView.ts]: This is what the elementDisopsable is for :) I will use that.

File: src/vs/workbench/browser/parts/views/customView.ts (hunk 3 above)
Comment (sandy081) [reply src/vs/workbench/browser/parts/views/customView.ts]: Is this because of using new tree?

File: src/vs/workbench/browser/parts/views/customView.ts [hunk 7]
```diff
@@ -530,136 +596,129 @@ export class CustomTreeView extends Disposable implements ITreeView {
...
-	renderTemplate(tree: ITree, templateId: string, container: HTMLElement): ITreeExplorerTemplateData {
+	renderTemplate(container: HTMLElement): ITreeExplorerTemplateData {
 		DOM.addClass(container, 'custom-view-tree-node-item');
 
 		const icon = DOM.append(container, DOM.$('.custom-view-tree-node-item-icon'));
+
 		const resourceLabel = this.labels.create(container, { supportHighlights: true, donotSupportOcticons: true });
-		DOM.addClass(resourceLabel.element, 'custom-view-tree-node-item-resourceLabel');
-		const actionsContainer = DOM.append(resourceLabel.element, DOM.$('.actions'));
+		const actionsContainer = DOM.append(container, DOM.$('.actions'));
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/customView.ts]: @alexr00 One minor question as asked before, is this change intentional? Why not use the same structure as before?

**Review by sandy081 — APPROVED** [review]

**Comment by sandy081** [comment]
@alexr00 I had just one last comment otherwise LGTM

File: src/vs/workbench/browser/parts/views/customView.ts (hunk 7 above)
Comment (alexr00) [reply src/vs/workbench/browser/parts/views/customView.ts]: It is needed to have the correct alignment of the actions. Appending to the `resourceLabel.element` causes the action buttons to be right up against the resource label, instead of nicely aligned to the right.

**Comment by alexr00** [comment]
Auto expand is still missing. I'm merging now to start getting selfhost bugs.
